lib/Common/Memory/CustomHeap.h

```diff
@@ -131,63 +131,100 @@ struct Allocation
 
 };
 
-/*
- * Simple free-listing based heap allocator
- *
- * Each allocation is tracked using a "HeapAllocation" record
- * Once we alloc, we start assigning chunks sliced from the end of a HeapAllocation
- * If we don't have enough to slice off, we push a new heap allocation record to the record stack, and try and assign from that
- */
-class Heap
+// Wrapper for the two HeapPageAllocator with and without the prereserved segment.
+// Supports multiple thread access. Require explicit locking (via CodePageAllocator::AutoLock)
+class CodePageAllocators
 {
 public:
-    Heap(AllocationPolicyManager * policyManager, ArenaAllocator * alloc, bool allocXdata);
-
-    Allocation* Alloc(size_t bytes, ushort pdataCount, ushort xdataSize, bool canAllocInPreReservedHeapPageSegment, bool isAnyJittedCode, _Inout_ bool* isAllJITCodeInPreReservedRegion);
-    bool Free(__in Allocation* allocation);
-    bool Decommit(__in Allocation* allocation);
-    void FreeAll();
-    bool IsInRange(__in void* address);
+    class AutoLock : public AutoCriticalSection
+    {
+    public:
+        AutoLock(CodePageAllocators * codePageAllocators) : AutoCriticalSection(&codePageAllocators->cs) {};
+    };
 
-    template<typename T>
-    HeapPageAllocator<T>* GetPageAllocator(Page * page)
+    CodePageAllocators(AllocationPolicyManager * policyManager, bool allocXdata, PreReservedVirtualAllocWrapper * virtualAllocator) :
+        pageAllocator(policyManager, allocXdata, true /*excludeGuardPages*/, nullptr),
+        preReservedHeapPageAllocator(policyManager, allocXdata, true /*excludeGuardPages*/, virtualAllocator),
+        cs(4000)
```

Comment:
Where does this spin count value come from?  Arbitrary?  Is it a reasonable count?